### PR TITLE
feat: improve guidance on PUT vs. POST as well other verbs.

### LINKED ISSUE #391: Better guidance for PUT vs. POST for creation
We currently don't give a clear guidance on the usage of HTTP methods for resource creation.
Currently our APIs have a wild mixture of both.

A naive POST has a problem in case of a connection issue that the client doesn't know whether it was successful, and retry might give a duplicated resource.
For a PUT you generally need to know the resource's ID before creation.

How can we make this clearer?

Some ideas:

* ID generation on client side, if possible (maybe UUID, but see [rule 144](https://opensource.zalando.com/restful-api-guidelines/#144)), then use PUT
* POST for ID generation, then PUT
* idempotent POST (using some foreign keys)
* POST with a way to find out whether it worked (GET)

### PR DESCRIPTION
**Pull request by tkrop**
Fixes #391.

I have revisited #391 and excluded the discussion we continued about `POST` vs. `PUT` vs. `PATCH` for update in the API Guild this week (see https://github.com/zalando/restful-api-guidelines/issues/391#issuecomment-403156420 and https://github.bus.zalan.do/Sokoban/rutherfordium/pull/17#discussion_r551990). So the changes are majorly editorial, removing some inconsistencies and improving the guidance on aspects we all commonly agreed about.

Unfortunately, the enhancements are a bit hidden between the reformatting. So please read carefully and make suggestions to improve wording. I have create a new [issue](#434) to track the remaining discussion.

### CODE REVIEW

File: chapters/http-requests.adoc [hunk 1]
```diff
@@ -7,198 +7,214 @@
...
+of <<182,`ETag` and `If-(None-)Match`>> headers should be considered to signal
+the server stricter demands to expose conflicts and prevent lost updates.
+The section <<optimistic-locking>> also describes some alternatives to this
+approach. 
+
+*Note:* In the rare cases where PUT is although used for resource creation,
+the resource IDs are maintained by the client and passed as a URL path segment.
+Putting the same resource twice is required to be idempotent and to result in
+the same single resource instance. If PUT is applied for creating a resource,
+only URIs should be allowed as resource IDs. If URIs are not available POST
```
Comment (whiskeysierra) [inline chapters/http-requests.adoc]: >  If PUT is applied for creating a resource, only URIs should be allowed as resource ID

I believe that will cause confusion. My first thought was this `/web-sites/http://example.org` (ignoring escaping). Have we specified the term resource ID somewhere? Most people refer to the last path segment not the whole URI.

File: chapters/http-requests.adoc [hunk 2]
```diff
@@ -7,198 +7,214 @@
...
 follows:
 
+
 [[get]]
 === GET
 
-GET requests are used to read a single resource or query set of
-resources.
+GET requests are used to *read* either a single resource or a set of resources
+on a list endpoint.
```
Comment (meshcalero) [inline chapters/http-requests.adoc]: not sure I like the distinction between a single resource or a list endpoint. For me a resource represents a specific concept. Whether this concept is a single entity or a concept of entities doesn't matter. I would use the term "set of resources" only for things like Multi-GET or embedded resources.

File: chapters/http-requests.adoc [hunk 3]
```diff
@@ -7,198 +7,214 @@
...
-same resource twice is by itself *not* required to be idempotent and may
-result in multiple resource instances. Anyhow, if external URIs are
-present that can be used to identify duplicate requests, it is best
-practice to implement POST in an idempotent way.
+POST requests are idiomatically used to *create* single resources on a
+collection resource endpoint, but other semantics on single resources endpoint
+are equally possible. The semantic for collection endpoints is best described
+as _"please add the enclosed representation to the collection resource
+identified by the URL"_. The semantic for single resource endpoints is best
+described as _"please execute the given well specified request on the resource
```
Comment (meshcalero) [inline chapters/http-requests.adoc]: > The semantic for single resource endpoints is best described as _"please execute the given well specified request on the resource
I would move this sentence below the two the list below, as what is stated in the list only applies to creates. For single resource POST we can not make any assumptions what it does and returns.

File: chapters/http-requests.adoc (hunk 2 above)
Comment (tkrop) [reply chapters/http-requests.adoc]: Okay, I change it back.

Comment (tkrop) [reply chapters/http-requests.adoc]: I wanted to say, I remove the term *list endpoint*.

> I would use the term "set of resources" only for things like Multi-GET or embedded resources.

Can you give me a hint, why *set of resources* should be not be used here to address the use case you mentioned here?

File: chapters/http-requests.adoc [hunk 4]
```diff
@@ -7,198 +7,214 @@
...
+* GET requests for collection resources may return either 200 (if the listing
+is empty) or 404 (if the list is missing)
+* GET requests must NOT have a request body payload
 
-* GET requests for individual resources will usually generate a 404 if
-the resource does not exist
-* GET requests for collection resources may return either 200 (if the
-listing is empty) or 404 (if the list is missing)
-* GET requests must NOT have request body payload
+*Note:* GET requests on collection resources should provide sufficient
```
[the remark this answers is not in the document]
Comment (tkrop) [reply chapters/http-requests.adoc]: `list` -> 27 (8 in `http-requests.adoc`)
`collection` -> 33 (16 in `http-requests.adoc`)

Any other opinions about this?

File: chapters/http-requests.adoc (hunk 1 above)
Comment (tkrop) [reply chapters/http-requests.adoc]: Good catch. I removed this at another location, but oversaw this usage.

File: chapters/http-requests.adoc (hunk 3 above)
Comment (tkrop) [reply chapters/http-requests.adoc]: Okay, I put it to the next paragraph below the list.

File: chapters/http-requests.adoc [hunk 5]
```diff
@@ -7,198 +7,214 @@
...
+by server and returned with response payload.
+
+*Tip:* Posting the same resource twice is by itself *not* required to be
+_idempotent_ and may result in multiple resource instances. However, it is
+best practice to implement POST _idempotent_, by temporary or permanently
+storing _foreign keys_ or _idempotency keys_ to identify re-send requests on
+the server. Here, a _foreign key_ is a natural partner resource ID, e.g. the
+shopping cart ID is a _foreign key_ candidate for an order, while an
+_idempotency key_ is an artificial request key that is preserved when
+re-sending the POST request.
```
[the remark this answers is not in the document]
Comment (tkrop) [reply chapters/http-requests.adoc]: Good point. I tried this by the example, but may forgot an abstract definition.

File: chapters/http-requests.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (tkrop) [reply chapters/http-requests.adoc]: The application of this request is easier than thought. I had only to switch 4 appearances from `list` to `collection`, which I think is more suitable.

**Comment by ddufourboivin-zalando** [comment]
👍

**Comment by meshcalero** [comment]
To me the first sentence on GET is still not fully appropriate

> GET requests are used to read either a single resource or a set resources

From my point of view a GET always reads a single resource. A resource represents some concept, which could be a single entity or a collection (usually of other resources).

I would just use the sentence "GET requests are used to read a resource", but if others prefer the current text, I wouldn't veto either.

**Comment by zeitlinger** [comment]
I'd write:

GET requests are used to read either a single or a collection resource.

.. because the following paragraphs refer to collection resources, not sets of resources.

**Comment by tkrop** [comment]
@meshcalero I took over @zeitlinger  suggestions. Do you approve?

**Comment by zeitlinger** [comment]
👍

**Comment by tkrop** [comment]
:+1:

**Comment by meshcalero** [comment]
:+1:

**Comment by tkrop** [comment]
:+1:

**Comment by zeitlinger** [comment]
:+1:

**Pull request merged**